lib/listen/fsm.rb

```diff
@@ -24,42 +27,54 @@ def states
         @states ||= {}
       end
 
-      # Declare an FSM state and optionally provide a callback block to fire
+      # Declare an FSM state and optionally provide a callback block to fire on state entry
       # Options:
+      # * start: make this the start state
       # * to: a state or array of states this state can transition to
-      def state(*args, &block)
-        if args.last.is_a? Hash
-          # Stringify keys :/
-          options = args.pop.each_with_object({}) { |(k, v), h| h[k.to_s] = v }
-        else
-          options = {}
-        end
-
+      def state(*args, start: nil, to: nil, &block)
         args.each do |name|
           name = name.to_sym
-          default_state name if options['default']
-          states[name] = State.new(name, options['to'], &block)
+          start_state(name) if start
+          states[name] = State.new(name, to, &block)
         end
       end
     end
 
     # Be kind and call super if you must redefine initialize
     def initialize
-      @state = self.class.default_state
+      @state = self.class.start_state
+      @mutex = ::Mutex.new
+      @state_changed = ::ConditionVariable.new
     end
 
-    # Obtain the current state of the FSM
+    # Current state of the FSM
     attr_reader :state
 
     def transition(state_name)
-      new_state = validate_and_sanitize_new_state(state_name)
-      return unless new_state
-      transition_with_callbacks!(new_state)
+      if (new_state = validate_and_sanitize_new_state(state_name))
+        transition_with_callbacks!(new_state)
+      end
     end
 
-    # Immediate state transition with no checks, or callbacks. "Dangerous!"
-    def transition!(state_name)
-      @state = state_name
+    # Low-level, immediate state transition with no checks or callbacks.
+    def transition!(new_state)
+      @mutex.synchronize do
+        yield if block_given?
+        @state = new_state
+        @state_changed.signal
```

Comment:
This Mutex#synchronize + `ConditionVariable#signal` is the guaranteed pattern for waking up code that is blocked on `wait` below.